[CK Tile] Stream K GEMM Kernel HostArgs and Kernel Classes

ecamartins opened the pull request:
## Proposed changes

As part of initial steps to integrate Stream K into CK Tile, this PR adds a two new classes: 1) `StreamKHostArgs` and 2) `StreamKKernel`. Similarly, to other CK Tile GEMM kernels, the StreamKKernel injects the `UniversalGemmKernel` to add necessary functionality for GEMM.

We are basing our initial Stream K implementation off old CK's example/01_gemm/gemm_xdl_streamk.cpp. Thus, we will be basing our TilePartitioner (coming in a future PR) off `BlockToCTileMap_GemmStreamK` in include/ck/tensor_operation/gpu/grid/block_to_ctile_map.hpp.

Similarly to Stream K in old CK, an instance of the `TilePartitioner` class is a kernel argument. The reason for this is because, for Stream K, old CK's `BlockToCTileMap_GemmStreamK` must be instantiated to determine important member variables like `sk_num_blocks` and `reduction_start_block_idx` which are required to determine the appropriate grid size and workspace buffer size. In other words, while the existing tile partitioners in CK Tile contain mostly static functions, this is not the case for Stream K's tile partitioner as most information is instance-specific.

## Checklist

Please put an `x` into the boxes that apply. You can also fill these out after creating the PR. If you're not sure, please don't hesitate to ask.

- [ ] I have added tests relevant to the introduced functionality, and the unit tests are passing locally
- [ ] I have added the test to REGRESSION_TESTS list defined at the top of CMakeLists.txt in tests/CMakeLists.txt, **IF** the test takes more than 30 seconds to run.
- [x] I have added inline documentation which enables the maintainers with understanding the motivation
- [ ] I have removed the stale documentation which is no longer relevant after this pull request
- [ ] (If this change is user-facing) I have added release notes which provide the end users with a brief summary of the improvement from this pull request
- [x] I have run `clang-format` on all changed files
- [ ] Any dependent changes have been merged

ecamartins committed:
CK Tile Stream K Device Ops

Implementation of CK Tile StreamKHostArgs and StreamKKernel classes. The
StreamKKernel class injects Universal Gemm and includes functions to
facilitate kernel preparation for the GPU.

ecamartins committed:
Merge branch 'develop' into emimarti/ck_tile/streamk_device_ops

ecamartins committed:
Stream K Device Ops Fixes

- Update GetWorkSpaceSize to call TilePartitioner's GetWorkSpaceSize to
  ensure we get size needed for accumulation buffers and semaphores.
- Pass in num_sk_blocks into TilePartitioner constructor
- Update documentation

AviralGoelAMD commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -0,0 +1,249 @@
+
+        return concat('_', "streamk", gemm_prec_str<ADataType, BDataType>(),
+                      concat('x', P_::MPerBlock, P_::NPerBlock, P_::KPerBlock),
+                      concat('x', P_::GetVectorSizeA(), P_::GetVectorSizeB(), P_::GetVectorSizeC()),
```
Could we also print WaveTile dimensions here? That would be helpful

ecamartins replied:
I agree. I have added the `WarpTile` dimensions in https://github.com/ROCm/composable_kernel/pull/2681/commits/fe8f3bbf64a019320cc647b99ff4f160df84e5ea.

AviralGoelAMD commented on include/ck_tile/ops/gemm.hpp:
```diff
@@ -1,5 +1,5 @@
-// SPDX-License-Identifier: MIT
-// Copyright (c) 2018-2025, Advanced Micro Devices, Inc. All rights reserved.
+// Copyright © Advanced Micro Devices, Inc., or its affiliates.
```
nit: For consistency, let us use (c) and not the symbol. 
reason: Because all the other files in CK use (c)

ecamartins replied:
Based on our discussion offline regarding documentation around copyright messages, it seems like the c-symbol is preferred in this case.

ecamartins committed:
Add WarpTile dimensions to GetName function in StreamKKernel class

ecamartins committed:
Merge branch 'develop' into emimarti/ck_tile/streamk_device_ops

AviralGoelAMD commented:
LGTM! I also appreciate the well written comments. ^_^

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -0,0 +1,251 @@
+    /// @brief For a given tile in the C tensor, one workgroup accumulates results of other
+    /// contributing workgroups
+    Reduction
+};
```
```suggestion
enum StreamKReductionStrategy : uint32_t
{
    /// @brief Workgroups atomically add their results to the C tensor
    Atomic = 0u,
    /// @brief For a given tile in the C tensor, one workgroup accumulates results of other
    /// contributing workgroups
    Reduction = 1u
};
```
Would recommend explicit datatype for enumeration.

Also for commenting, there seems to be a mix of inline (/// ) and block (/* */) style in most of CK code. Do you know if there is any consensus for CK style? :)

ecamartins replied:
Thanks for the feedback! It is addressed below:

> Would recommend explicit datatype for enumeration.

I added an explicit datatype for the enum in https://github.com/ROCm/composable_kernel/pull/2681/commits/cda189b1608aaa5f82cced5f3475fcc5c7d9a30c.

>  Do you know if there is any consensus for CK style? 

Looking at the other kernel files, there seems to be some mixing and matching between the two styles with inline being more prominent. So, I have unified the commenting style to use the inline style in https://github.com/ROCm/composable_kernel/pull/2681/commits/89f77c1675c16f5cd7a556d2daf2ed354a53a8ae.

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -0,0 +1,251 @@
+///      arguments object. It contain all necessary information required to build proper kernel
+///      argument and launch kernel on GPU. This structure defines the GEMM problem configuration by
+///      stating all required information like M,N,K sizes and respective strides.
+struct StreamKHostArgs : public ck_tile::UniversalGemmHostArgs<>
```
```suggestion
///      This structure is passed to @ref StreamKKernel "StreamKKernel" when creating the kernel
///      arguments object. It contains all necessary information required to build proper kernel
///      arguments and launch the kernel on GPU. This structure defines the GEMM problem configuration by
///      stating all required information like M,N,K sizes and respective strides.
struct StreamKHostArgs : public ck_tile::UniversalGemmHostArgs<>
```

cgmillette commented on include/ck_tile/ops/gemm/kernel/streamk_gemm_kernel.hpp:
```diff
@@ -0,0 +1,251 @@
+    CK_TILE_HOST static StreamKKernelArgs MakeKernelArgs(const StreamKHostArgs& host_args)
+    {
+        int occupancy = StreamKKernel::Occupancy();
+        int num_cu    = StreamKKernel::NumCU();
```
```suggestion
        int occupancy = Occupancy();
        int num_cu    = NumCU();
```
Don't need class scoping here

ecamartins replied:
Thanks for pointing this out.

I have addressed your feedback in https://github.com/ROCm/composable_kernel/pull/2681/commits/19f48ea938613a10bf2660a84f3a22d9af22cda2. I also removed some unnecessary scoping in some other spots.

cgmillette commented:
Great first look at the StreamK invoker!

ecamartins committed:
Fix typos in StreamKHostArgs class description.

Co-authored-by: Christopher Millette <63608002+cgmillette@users.noreply.github.com>

ecamartins committed:
Apply clang format on updated comment for StreamKHostArgs

ecamartins committed:
Explicitly specify type for StreamKReductionStrategy enum

ecamartins committed:
Remove unecessary scopes

ecamartins committed:
Unify the commenting style to inline comments

ecamartins committed:
Add explicit casts for occupancy and num_cu in MakeKernelArgs function

Both the static functions Occupancy and NumCU in the StreamKKernel class
use functions from the HIP API that result in the returned occupancy and
num_cu types being type int. The TilePartitioner interface for stream K will
have occupancy and num_cu being type ck_tile::index_t which is int32_t.
Thus, to be safe, this change ensures that both occupancy and num_cu are
cast to int32_t.

ecamartins committed:
Merge branch 'develop' into emimarti/ck_tile/streamk_device_ops

ecamartins committed:
Fix use of kentry due to interface update
PR #2594 updated the interface for the kentry function in
include/ck_tile/host/kernel_launch.hpp. As a result, the static function
Occupancy was updated to work correctly with the new interface.
PR #2594 also changed UniversalGemmKernel's KernelBlockSize static
variable to kBlockSize, so the StreamKKernel class was updated to
reflect this change.

ecamartins committed:
Switch type of num_sk_blocks from uint32_t to int32_t

This change switches the type of num_sk_blocks to type ck_tile::index_t
which is int32_t. This was done because parallel work for the CK Tile
StreamK TilePartitioner's constructor will have num_sk_blocks as
ck_tile::index_t. Thus, this change will help unify the interfaces to
avoid any type conversion errors.

ecamartins committed:
Merge branch 'develop' into emimarti/ck_tile/streamk_device_ops

ecamartins committed:
Merge branch 'develop' into emimarti/ck_tile/streamk_device_ops

cgmillette reviewed (approved):
LGTM!
